module/apmgoredis/integration_test.go

```diff
@@ -0,0 +1,391 @@
+		}
+
+		defer client.Close()
+		cleanRedis(t, client, testCase.isCluster)
+
+		_, spans, _ := apmtest.WithTransaction(func(ctx context.Context) {
+			wrappedClient := apmgoredis.Wrap(client).WithContext(ctx)
+
+			var incr1 *redis.IntCmd
+			var incr2 *redis.IntCmd
+			var incr3 *redis.IntCmd
+			_, err := wrappedClient.TxPipelined(func(pipe redis.Pipeliner) error {
+				incr1 = pipe.Incr("foo")
+				assert.NoError(t, incr1.Err())
+
+				incr2 = pipe.Incr("bar")
+				assert.NoError(t, incr2.Err())
+
+				incr3 = pipe.Incr("bar")
+				assert.NoError(t, incr3.Err())
+
+				return nil
+			})
+
+			assert.Equal(t, int64(1), incr1.Val())
+			assert.Equal(t, int64(1), incr2.Val())
+			assert.Equal(t, int64(2), incr3.Val())
+
+			assert.NoError(t, err)
+		})
+
+		switch testCase.isTxWrapped {
+		case true:
+			assert.Len(t, spans, 1)
+			assert.Equal(t, "(pipeline) INCR INCR INCR", spans[0].Name)
+		case false:
+			assert.Len(t, spans, 0)
+		}
+	}
+}
+
+func TestPipelineTransaction(t *testing.T) {
+	testCases := getTestCases(t)
+
+	for _, testCase := range testCases {
+		client := testCase.client
+		if client == nil {
+			t.Errorf("cannot create client")
+		}
+
+		defer client.Close()
+		cleanRedis(t, client, testCase.isCluster)
+
+		_, spans, _ := apmtest.WithTransaction(func(ctx context.Context) {
+			wrappedClient := apmgoredis.Wrap(client).WithContext(ctx)
+
+			pipe := wrappedClient.TxPipeline()
+
+			incr1 := pipe.Incr("foo")
+			assert.NoError(t, incr1.Err())
+
+			incr2 := pipe.Incr("bar")
+			assert.NoError(t, incr2.Err())
+
+			incr3 := pipe.Incr("bar")
+			assert.NoError(t, incr3.Err())
+
+			_, err := pipe.Exec()
+			require.NoError(t, err)
+
+			assert.Equal(t, int64(1), incr1.Val())
+			assert.Equal(t, int64(1), incr2.Val())
+			assert.Equal(t, int64(2), incr3.Val())
+		})
+
+		switch testCase.isTxWrapped {
+		case true:
+			assert.Len(t, spans, 1)
+			assert.Equal(t, "(pipeline) INCR INCR INCR", spans[0].Name)
+		case false:
+			assert.Len(t, spans, 0)
+		}
+	}
+}
+
+func redisClient(t *testing.T) *redis.Client {
+	redisURL := os.Getenv("GOREDIS_URL")
+	if redisURL == "" {
+		t.Skipf("GOREDIS_URL not specified")
+	}
+
+	var err error
+	redisURL, err = lookupIPFromHostPort(redisURL)
+	if err != nil {
+		return nil
+	}
+
+	closeConn := true
+	client := redis.NewClient(&redis.Options{
+		Addr: redisURL,
+	})
+
+	defer func() {
+		if closeConn {
+			client.Close()
+		}
+	}()
+
+	closeConn = false
+	return client
+}
+
+func redisClusterClient(t *testing.T) *redis.ClusterClient {
+	redisURLs := strings.Split(os.Getenv("GOREDIS_CLUSTER_URLS"), " ")
+	if len(redisURLs) == 0 {
+		if t != nil {
+			t.Skipf("GOREDIS_CLUSTER_URLS not specified")
+		}
+	}
+
+	for i, redisURL := range redisURLs {
+		var err error
+		redisURLs[i], err = lookupIPFromHostPort(redisURL)
+		if err != nil {
+			return nil
+		}
+	}
+
+	closeConn := true
+	client := redis.NewClusterClient(&redis.ClusterOptions{
+		Addrs: redisURLs,
+	})
+
+	defer func() {
+		if closeConn {
```

Comment:
same as in redisClient: lose the closeConn and defer, as this will never be reached.